Remove sync constellation -> embedder communication

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #22042 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22121)
<!-- Reviewable:end -->

gterzian committed:
constellation: restructure navigation, remove sync comm

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/constellation.rs
 * @cbrewster: components/constellation/constellation.rs
 * @paulrouget: ports/servo/browser.rs, components/constellation/constellation.rs, components/compositing/windowing.rs, components/servo/lib.rs
 * @kichjang: components/script_traits/lib.rs

gterzian commented:
@jdm this is the only case of `recv()` involving a channel send inside an `EmbedderMsg` I could find in the constellation(there are some other cases involving canvas and layout however).

In the case of the unloading of a document, or the opening of an auxiliary, it's the script-thread that sends the `EmbedderMsg`(via a forwarding by the constellation), and blocks waiting on the reply from the embedder...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/AmY_sXF9QVCxeJwbjMkylg)

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 629fe8e47bb64c8efd3619139d42a7710975d276 with merge 4bc52010045a41e483569b21081c568a39175262...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/ElU2P4sjQOqRk5Qo0kDcfw)

gterzian commented:
```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/unloading-documents/unload/002.html", "line": 105047, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html", "line": 105249, "action": "test_result", "expected": "OK"}
{"status": "ERROR", "group": "default", "message": "Got results from /html/browsers/browsing-the-web/history-traversal/support/window-name-test.sub.html?shouldhavename=true&sendmessage=true, expected /html/browsers/browsing-the-web/history-traversal/window-name-after-cross-origin-sub-frame-navigation.sub.html\nTraceback (most recent call last):\n  File \"/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py\", line 162, in run_test\n    result = self.do_test(test)\n  File \"/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservo.py\", line 110, in do_test\n    result = self.convert_result(test, self.result_data)\n  File \"/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py\", line 69, in __call__\n    (result_url, test.url))\nAssertionError: Got results from /html/browsers/browsing-the-web/history-traversal/support/window-name-test.sub.html?shouldhavename=true&sendmessage=true, expected /html/browsers/browsing-the-web/history-traversal/window-name-after-cross-origin-sub-frame-navigation.sub.html\n", "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/history-traversal/window-name-after-cross-origin-sub-frame-navigation.sub.html", "line": 105297, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/forms/form-submission-0/getactionurl.html", "line": 105537, "action": "test_result", "expected": "OK"}
```

Given all those timeouts, and more, it looks likes something is wrong...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 610bbe3f10a96c0cfddeb1b6a4e700a52776ac0f with merge a38dfc29908cf97c889d87d6579512b086b273a3...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10431)

gterzian commented:
@bors-servo try (seeing if immediate handling of the message and window event prevents those timeouts)

bors-servo commented:
:hourglass: Trying commit 5ea17fbb16867af3ced2988a76c252e203c77fdc with merge d829e4b111c76b37e1be49b8f4319bca84d0cee8...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/NKyz8Ip5Sxel1O6ZaoYxoA)

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/ZbS1fXwYT7C-L-Cqi8Xi5w)

paulrouget commented:
Can you give some more context about what motivates this change.

Can we run into some ordering issue if multiple inner frames try to navigate at the same time? Don't we want to pass the pipeline id instead of the top level browsing context id to avoid a mix up?

gterzian commented:
@paulrouget The main goal of https://github.com/servo/servo/issues/22042 seems to be removing this `recv()` at https://github.com/servo/servo/pull/22121/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fL2067

For other communication with the embedder, like opening pop-ups with `window.open`, and prompting in `before_unload`, we are already sending the sender directly to the embedder(via a forwarding of the message by the constellation) and it's the script content that blocks waiting for the answer(which seems spec compliant).

You were right on the `PipelineId`, actually not including it in the message was what was causing the timeouts in the case of iframe loads...

The timeouts seems ok now. I applied `rustfmt` since CI was complaining, however that also gave me a bunch of unrelated changes, so I put it in a separate commit...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 29666a84f218d578a93d0d21cdcb4a63192cfb3d with merge f864f44e09b3dcaca1f1cb702e73396693ba44a6...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Uu5p0qZ3Sm-y4OhFs9X36Q)

gterzian commented:
```
Incorrect sha1 hash for /repo/android-toolchains/android-ndk-r12b-linux-x86_64.zip
[taskcluster 2018-11-10 05:03:31.981Z] === Task Finished ===
[taskcluster 2018-11-10 05:03:32.098Z] Artifact "public/servoview.aar" not found at "/repo/target/i686-linux-android/release/servoview.aar"
[taskcluster 2018-11-10 05:03:32.210Z] Artifact "public/servoapp.apk" not found at "/repo/target/i686-linux-android/release/servoapp.apk"
[taskcluster 2018-11-10 05:03:33.281Z] Unsuccessful task run with exit code: 1 completed in 86.258 seconds

```

gterzian commented:
`JSON.parse: unexpected end of data at line 1 column 1 of the JSON data`
I think given all these errors there must still be a problem with the roundtrip constellation -> embedder, maybe the hash key needs to be more precise...

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -2084,14 +2120,33 @@ where
     ) {
-        self.load_url(top_level_browsing_context_id, source_id, load_data, replace);
+        let key = (
+            top_level_browsing_context_id,
```
I don't think you need the top level browsing context part of the key. The pipeline id will always be more precise.

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -2084,14 +2120,33 @@ where
+            .pending_approval_navigations
+            .entry(key)
+            .or_insert(VecDeque::new());
+        pending.push_back((load_data.clone(), replace));
```
Is it possible to have multiple pending navigation for one pipeline? I think it has to be unique. If there's already a pending navigation, fail.

paulrouget commented on components/script_traits/lib.rs:
```diff
@@ -720,6 +720,8 @@ pub enum ConstellationMsg {
     /// Inform the constellation of a key event.
     Keyboard(KeyboardEvent),
+    /// Whether to allow script to navigate.
+    AllowNavigation(TopLevelBrowsingContextId, PipelineId, ServoUrl, bool),
```
Just to make things easier to understand, can you rename `AllowNavigation` to `AllowNavigationResponse`?

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -992,6 +999,40 @@ where
+                allowed,
+            ) => {
+                if !allowed {
+                    return;
```
Don't you need to remove the pending navigation from pending_approval_navigations?

gterzian commented:
Thanks, I addressed your comments.

I'm still debugging an issue that gives me a `JSON.parse: unexpected end of data at line 1 column 1 of the JSON data`  at https://github.com/servo/servo/blob/b1a2b6b5bfe72440059834c67dc28ad422b89a82/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js#L249

The problem only seem to arise in the context of referrer policy tests.

I've made some sanity checks and it is really caused by the additional roundtrip, since removing the change makes the test pass.

I've looked into the fact that we're now storing the `LoadData`, which creates new references of `url` and `referrer_url`, which are actually `Arc<Url>`, and `Arc::make_mut` is called on the referrer url, which is mutated as part of the referrer policy workflow. I though maybe that the extra reference stored by the constellation cause unexpected problems since that would result in `Arc::make_mut` cloning the data, but that seems like a dead-end(Also, the referrer stuff would only be done after the navigation has been approved, it seems). 

It's also weird that the error arises with an `xhr`, because those request don't seem to actually go throught the "embedder approval of the navigation" flow. 

Anyway, still looking into it...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22142) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented:
@paulrouget Ok, it's ready for another round of review...

paulrouget reviewed (commented):
Looks good to me, but you need to make the message and the hashmap.

There's only one pending navigation per pipeline (so drop the url from the hashmap) and transporting the url and the top level browsing context is not necessary (so remove them from response message).

gterzian commented:
@paulrouget thanks, comments addressed!

paulrouget commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f7fc1dd has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget f7fc1dd390c7e4e13b13f0b742f883573adf7c59 -->

bors-servo commented:
:hourglass: Testing commit f7fc1dd390c7e4e13b13f0b742f883573adf7c59 with merge 0b28050d451a93bc8de24b41daf3183075e8dd3e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10956)

gterzian commented:
Ok so the problem hadn't gone away after all(I forgot it was in the "mozilla" test suite...).

Time to add some print statements...

gterzian commented:
Ok so it looks like the problem relates to the `load` event of the iframe(the test expects a subresource to have loaded by the time the iframe fires the event https://github.com/servo/servo/blob/a04200ae795d6f177ed77e9f826a8151f48d78ec/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js#L235). 

I think it actually can be traced back to the concept of [#delaying-load-events-mode](https://html.spec.whatwg.org/multipage/browsers.html#delaying-load-events-mode), which is still unimplemented.

I presume the blocking "asking for permission to the embedder" call actually "by chance" functioned as a load blocker(the iframe load event is fired in response to a message from the constellation, itself sent in response to handling the message sent here: https://html.spec.whatwg.org/multipage/browsers.html#delaying-load-events-mode). 

So there's a lot of back and forth, and I still haven't figured it out yet...

gterzian commented:
@jdm I saw the `LoadBlocker` used by the `HTMLIFrameElement`, and what I'm trying to figure out is how this relates to this part of the spec:

> A browsing context that is a nested browsing context can be put into a delaying load events mode. This is used when it is navigated, to delay the load event of its browsing context container before the new Document is created. (https://html.spec.whatwg.org/multipage/browsers.html#delaying-load-events-mode)

and also how this relates to the concept of [delaying-the-load-event](https://html.spec.whatwg.org/multipage/parsing.html#delay-the-load-event) of a document. 

When you look at https://html.spec.whatwg.org/multipage/parsing.html#the-end, you can see:

- Step 6 "Spin the event loop until there is nothing that delays the load event in the Document."(which would include a nested BC in "delaying load events mode")
- Step 7 "Queue a task to  fire an event named load"
- Step 12 "Queue a task to mark the Document as completely loaded."

Then in [#iframe-load-event-steps](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#iframe-load-event-steps), we can read:

> When a Document in an iframe is marked as completely loaded, the user agent must run the iframe load event steps. (Which includes "Fire an event named load at the iframe element.")


So the "completely loaded" part of a document is currently not implemented, although perhaps this could be said to be it: https://github.com/servo/servo/blob/a04200ae795d6f177ed77e9f826a8151f48d78ec/components/script/dom/document.rs#L1949

And I'm also wondering if there is a difference between "delaying the load of a document using load blockers" and "delaying the load of an iframe when the nested bc is being navigated". 

Finally, the reason why I am looking into this, is because the tests that started failing seems to relate to the "load" event of an iframe being fired before it has finished navigating, and previously the constellation would block waiting for a reply from the embedder, and then immediately start the navigation, while with the current change the "keeps running" while waiting for the reply from the embedder, and it seems that this "keeping running" results in the "load" event of the iframe being fired, while the navigation hasn't started yet(constellation still hasn't handled the reply from the embedder yet). 

So essentially, removing the blocking in the constellation seems to have opened up a timing issue, and I think this relates to the "load" event for the iframe not being delayed "as long as it should"...

gterzian commented:
The currently failing test(https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/referrer-policy/no-referrer/http-rp/cross-origin/http-http/a-tag/generic.keep-origin-redirect.http.html) seems to be doing the following:

1. Load the page at generic.keep-origin-redirect.http.htm
2. Create an iframe whose `src` will load https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html
3. a-tag.html navigates the iframe at https://github.com/servo/servo/blob/af1ced3148384c19424e004f054bfd648515b384/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html#L43
4. When the iframe fires the load event at https://github.com/servo/servo/blob/af1ced3148384c19424e004f054bfd648515b384/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js#L235, the test code expects "document.py" to already be in the "stash". Currently it's not, so we get the error of "empty json response" at https://github.com/servo/servo/blob/af1ced3148384c19424e004f054bfd648515b384/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js#L249

gterzian commented:
Ok so here is what is happening:

1. iframe is navigated -> navigation message sent to constellation.
2. the document of the iframe loads (I assume 1 happens just before in the script event-loop, since `maybe_queue_document_load` happens after running a task in the script event-loop) -> `ScriptMsg::LoadComplete` sent to constellation.
3. Constellation handles the navigation message. 
4. Constellation handles `LoadComplete`, and since it's an iframe sends a 'DispatchIFrameLoadEvent` message to scriot.
5. script handles messages, fires "load" on iframe. 
6. Constellation handles `AllowNavigation` reply from embedder, start the load for the iframe. 

Previously, the constellation would block at step 3, and then immediatelly start loading the url. Now, it sends a message to the embedder without blocking, allowing it to handle the `ScriptMsg::LoadComplete` in the meatime, and send a 'DispatchIFrameLoadEvent` back to script. 

I think this might ty into this [#delaying-load-events-mode](https://html.spec.whatwg.org/multipage/browsers.html#delaying-load-events-mode), which appears to be a different concept from "delaying the load event of a document(iframe or not)". 

The spec says we should put the nested browsing context  in "delaying load events mode" when navigation starts, and [take it out when the navigation matures](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents:delaying-load-events-mode-2). 

I think that might actually be the solution, and we were previously unaware of this problem because the blocking "allow navigation" call to the embedder would prevent the load event from firing before the load started.

gterzian committed:
implement windowproxy "delay-load-event-mode", and partially document "completely-loaded"

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 34755391a96f7410f8a6c1eeb88331b53caab162 with merge 113f7359fb1c98ae2535c16ac4b5964717efe9d8...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/cX0ysNMWQGuF5gz9nSoBtA)

gterzian commented:
Ok so there's something I still need to check with regards to that crash...

gterzian committed:
check for a document being "fully-active" before reflows

gterzian commented:
Ok so here is the problem:

1. We have an "active" iframe, whose parent is "inactive". Not "active" is not "fully active". 
2. The Layout thread for the "active" iframe has already quit. (See `ScriptThread.handle_exit_pipeline_msg`).
3. The "active" but not "fully-active" iframe is kept alive by a reference inside a script. 
4. `Document.Close` is called on that iframe inside that script. 
5. This leads to `Document.finish_load` being called, which sends a `Reflow` message to the layout thread, which has already exited by now. 
6. Panic. 

So the question is: do we want to not close the layout thread, or even not close the pipeline, for a document that is "active" yet not "fully active"? Or do we want to add a check inside `finish_load` that does not issue the reflow if the document is not "fully active"?

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/14544.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#14544.

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 98270969515b9424c9ac0ec7cfc666f83fc74e8f with merge b993c4971801bf9900f90e82d0bb56ba2df69422...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/I8azkEneR_ymLWUgqtwT2g)

gterzian commented:
```
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/context/context-release-with-workers.html", 
    "line": 56799, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/context/context-release-with-workers.html", 
    "line": 56860, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/ogles/GL/faceforward/faceforward_001_to_006.html", 
    "line": 57710, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/ogles/GL/faceforward/faceforward_001_to_006.html", 
    "line": 57723, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/semantics/scripting-1/the-script-element/async_010.htm", 
    "line": 58114, 
    "action": "test_result", 
    "expected": "OK"
}
```

Looks intermittent, passes locally...

gterzian commented:
@paulrouget  Ready for review...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/OQZBkHp0S9qAWbGR8ZCtfA)

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -1042,6 +1049,51 @@ where
+                                };
+                            if let Some(_parent_pipeline_id) = parent_pipeline_id {
+                                self.handle_subframe_loaded(pipeline_id);
+                            }
```
If the comment is correct, as in "concerns an iframe", I'm not sure about the matching logic. What exactly are you testing here? If the pipeline is part of a browsing context that is a top level browsing context?

gterzian replied:
Ah yes, you're right, I've updated this to check whether the pipeline belongs to a top-level BC or not...

paulrouget commented:
This looks good to me. I'd like to get @jdm's feedback on the changes under components/script/.

@jdm mind taking a look?

gterzian commented on components/constellation/constellation.rs:
```diff
@@ -1068,6 +1068,32 @@ where
+                            // we need to take it out of it's "delaying-load-events-mode".
+                            // https://html.spec.whatwg.org/multipage/#delaying-load-events-mode
+                            if !pipeline_is_top_level_pipeline {
+                                let msg = ConstellationControlMsg::StopDelayingLoadEventsMode(
```
@paulrouget I've changed this to sending a dedicated message, as opposed to re-use `handle_subframe_loaded`, since I think that would have resulted in double load events for the iframe...

jdm commented on components/script/script_thread.rs:
```diff
@@ -2166,6 +2176,20 @@ impl ScriptThread {
+                            let window_proxy = window.window_proxy();
+                            // https://html.spec.whatwg.org/multipage/
+                            // #navigating-across-documents:delaying-load-events-mode-2
+                            if let Some(_parent) = window_proxy.parent() {
```
This can use `.is_some()` instead if `if let`.

jdm commented on components/script/dom/document.rs:
```diff
@@ -1735,7 +1735,7 @@ impl Document {
             },
             LoadType::PageSource(_) => {
-                if self.has_browsing_context {
+                if self.has_browsing_context && self.is_fully_active() {
```
Which part of the spec does this check come from?

gterzian replied:
I think both are actually not spec compliant, yet necessary as bandaid because of the fact that layout will have already exited and we'll panic if we continue here if the document is not "fully active". 

I have filed https://github.com/servo/servo/issues/22507 to investigate the underlying issue...

What do you think?

jdm commented on components/script/dom/document.rs:
```diff
@@ -1900,6 +1900,7 @@ impl Document {
     // https://html.spec.whatwg.org/multipage/#the-end
     pub fn maybe_queue_document_completion(&self) {
         let not_ready_for_load = self.loader.borrow().is_blocked() ||
+            !self.is_fully_active() ||
```
What part of the spec is this implementing?

jdm commented on components/script/script_thread.rs:
```diff
@@ -1314,6 +1314,10 @@ impl ScriptThread {
         // into this loop too, but for now it's only images.
         debug!("Issuing batched reflows.");
         for (_, document) in self.documents.borrow().iter() {
+            // https://html.spec.whatwg.org/multipage/#event-loop-processing-model:fully-active-9
```
This can just be `// Step 13`.

jdm reviewed (changes requested):
I had not noticed the part of the spec that talks about a window proxy delaying the load event; good catch! Impressive investigations while writing this patch, as well. I did not expect solving this issue to be this complicated!

gterzian commented on components/script/dom/windowproxy.rs:
```diff
@@ -314,6 +318,26 @@ impl WindowProxy {
+    pub fn stop_delaying_load_events_mode(&self) {
+        self.delaying_load_events_mode.set(false);
+        if let Some(document) = self.document() {
+            if !document.loader().events_inhibited() {
```
Added this check, otherwise it would fail at https://github.com/servo/servo/blob/f3ca48206e02d48fc8e9f7d340d25e2210d00723/components/script/dom/document.rs#L1907

gterzian replied:
If there is still a loader active, the document will be queued for completion later...

gterzian commented:
squashed two commits together which actually wouldn't have compiled separately...

paulrouget commented on components/script/script_thread.rs:
```diff
@@ -2710,6 +2739,13 @@ impl ScriptThread {
             incomplete.parent_info,
             incomplete.opener,
         );
+        if let Some(_parent) = window_proxy.parent() {
```
`is_some()`

gterzian commented:
@paulrouget comments addressed...

paulrouget commented:
rustfmt issues.

gterzian commented:
Ok, should be fixed...

paulrouget commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit b7c1b9b has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget b7c1b9bc37e433eb66bc0bca3373efbd374ec6a5 -->

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11229)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit b7c1b9bc37e433eb66bc0bca3373efbd374ec6a5 with merge 2889315f14f4423e223558118ab3f204d4eb28ca...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11294)

jdm commented:
@bors-servo retry
* #13480

jdm commented:
```
 ▶ TIMEOUT [expected ERROR] /html/semantics/scripting-1/the-script-element/execution-timing/083.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ Navigated pipeline doesn't have a top_level_browsing_context (thread Constellation, at libcore/option.rs:1010)
  │ [2018-12-22T16:53:51Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/083.html:1:76 top is null
  │ stack backtrace:
  │    0:     0x7fb187d91ebc - backtrace::backtrace::trace::h77d04c33b09dac10
  │    1:     0x7fb187d910e2 - <backtrace::capture::Backtrace as core::default::Default>::default::hd8f68beaec462ca3
  │    2:     0x7fb187d91158 - backtrace::capture::Backtrace::new::hc8cf18d80aa809da
  │    3:     0x7fb185626979 - servo::main::{{closure}}::h5f924ab4bf93a4ed
  │    4:     0x7fb188a693a3 - std::panicking::rust_panic_with_hook::hd470ff3b3e7a1b18
  │                         at libstd/panicking.rs:480
  │    5:     0x7fb188a68f09 - std::panicking::continue_panic_fmt::hf0cf39ae1e114602
  │                         at libstd/panicking.rs:390
  │    6:     0x7fb188a68e05 - rust_begin_unwind
  │                         at libstd/panicking.rs:325
  │    7:     0x7fb188a8a56b - core::panicking::panic_fmt::hf53111ba26d35bb0
  │                         at libcore/panicking.rs:77
  │    8:     0x7fb188a8a5e1 - core::option::expect_failed::h3d6fec7eaa423a29
  │                         at libcore/option.rs:1010
  │    9:     0x7fb185783499 - <constellation::constellation::Constellation<Message, LTF, STF>>::handle_request_from_compositor::h181c14d921c1d7c9
  │   10:     0x7fb185788137 - <constellation::constellation::Constellation<Message, LTF, STF>>::run::h41a0b985492c71bd
  │   11:     0x7fb185650593 - std::sys_common::backtrace::__rust_begin_short_backtrace::hb5915c9409353fc7
  │   12:     0x7fb1856bfa87 - _ZN3std9panicking3try7do_call17hddb5f3f60d0085dcE.llvm.16664199431785739934
  │   13:     0x7fb188a7b619 - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:102
  │   14:     0x7fb185677b00 - <F as alloc::boxed::FnBox<A>>::call_box::h8a34e0350d11a190
  │   15:     0x7fb188a6628a - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$RP$$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h9d1cd498499ed135
  │                         at liballoc/boxed.rs:682
  │                          - std::sys_common::thread::start_thread::ha4bfe77455e6e434
  │                         at libstd/sys_common/thread.rs:24
  │   16:     0x7fb188a52e55 - std::sys::unix::thread::Thread::new::thread_start::hebe133da3ea5230a
  │                         at libstd/sys/unix/thread.rs:90
  │   17:     0x7fb18328a183 - start_thread
  │   18:     0x7fb181b5103c - clone
  │   19:                0x0 - <unknown>
  └ [2018-12-22T16:53:51Z ERROR servo] Navigated pipeline doesn't have a top_level_browsing_context
```
This might be related.

bors-servo commented:
:hourglass: Testing commit b7c1b9bc37e433eb66bc0bca3373efbd374ec6a5 with merge ce9bdc165769e463cd699beabfc2c88f3a616250...

jdm commented:
Yeah, this time it was:
```
  ▶ TIMEOUT [expected ERROR] /html/semantics/scripting-1/the-script-element/execution-timing/083.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ Navigated pipeline doesn't have a top_level_browsing_context (thread Constellation, at libcore/option.rs:1010)
  │ [2018-12-22T18:23:33Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/083.html:1:76 top is null
  │ stack backtrace:
  │    0:     0x7f82c7b568fc - backtrace::backtrace::trace::h77d04c33b09dac10
  │    1:     0x7f82c7b55b22 - <backtrace::capture::Backtrace as core::default::Default>::default::hd8f68beaec462ca3
  │    2:     0x7f82c7b55b98 - backtrace::capture::Backtrace::new::hc8cf18d80aa809da
  │    3:     0x7f82c53e7bc9 - servo::main::{{closure}}::h5f924ab4bf93a4ed
  │    4:     0x7f82c882dde3 - std::panicking::rust_panic_with_hook::hd470ff3b3e7a1b18
  │                         at libstd/panicking.rs:480
  │    5:     0x7f82c882d949 - std::panicking::continue_panic_fmt::hf0cf39ae1e114602
  │                         at libstd/panicking.rs:390
  │    6:     0x7f82c882d845 - rust_begin_unwind
  │                         at libstd/panicking.rs:325
  │    7:     0x7f82c884efab - core::panicking::panic_fmt::hf53111ba26d35bb0
  │                         at libcore/panicking.rs:77
  │    8:     0x7f82c884f021 - core::option::expect_failed::h3d6fec7eaa423a29
  │                         at libcore/option.rs:1010
  │    9:     0x7f82c55446e9 - <constellation::constellation::Constellation<Message, LTF, STF>>::handle_request_from_compositor::h181c14d921c1d7c9
  │   10:     0x7f82c5549387 - <constellation::constellation::Constellation<Message, LTF, STF>>::run::h41a0b985492c71bd
  │   11:     0x7f82c54117e3 - std::sys_common::backtrace::__rust_begin_short_backtrace::hb5915c9409353fc7
  │   12:     0x7f82c5480cd7 - _ZN3std9panicking3try7do_call17hddb5f3f60d0085dcE.llvm.16664199431785739934
  │   13:     0x7f82c8840059 - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:102
  │   14:     0x7f82c5438d50 - <F as alloc::boxed::FnBox<A>>::call_box::h8a34e0350d11a190
  │   15:     0x7f82c882acca - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$RP$$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h9d1cd498499ed135
  │                         at liballoc/boxed.rs:682
  │                          - std::sys_common::thread::start_thread::ha4bfe77455e6e434
  │                         at libstd/sys_common/thread.rs:24
  │   16:     0x7f82c8817895 - std::sys::unix::thread::Thread::new::thread_start::hebe133da3ea5230a
  │                         at libstd/sys/unix/thread.rs:90
  │   17:     0x7f82c304c183 - start_thread
  │   18:     0x7f82c191303c - clone
  │   19:                0x0 - <unknown>
  └ [2018-12-22T18:23:33Z ERROR servo] Navigated pipeline doesn't have a top_level_browsing_context

  ▶ CRASH [expected OK] /html/semantics/scripting-1/the-script-element/execution-timing/084.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2018-12-22T18:24:01Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/084.html:1:76 top is null
  │ [2018-12-22T18:24:01Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/resources/testharness.js:1477:19 Tried to create a test with file_is_test
  │ called `Option::unwrap()` on a `None` value (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at libcore/option.rs:355)
  │ stack backtrace:
  │    0:     0x7f0c199ae8fc - backtrace::backtrace::trace::h77d04c33b09dac10
  │    1:     0x7f0c199adb22 - <backtrace::capture::Backtrace as core::default::Default>::default::hd8f68beaec462ca3
  │    2:     0x7f0c199adb98 - backtrace::capture::Backtrace::new::hc8cf18d80aa809da
  │    3:     0x7f0c1723fbc9 - servo::main::{{closure}}::h5f924ab4bf93a4ed
  │    4:     0x7f0c1a685de3 - std::panicking::rust_panic_with_hook::hd470ff3b3e7a1b18
  │                         at libstd/panicking.rs:480
  │    5:     0x7f0c1a685949 - std::panicking::continue_panic_fmt::hf0cf39ae1e114602
  │                         at libstd/panicking.rs:390
  │    6:     0x7f0c1a685845 - rust_begin_unwind
  │                         at libstd/panicking.rs:325
  │    7:     0x7f0c1a6a6fab - core::panicking::panic_fmt::hf53111ba26d35bb0
  │                         at libcore/panicking.rs:77
  │    8:     0x7f0c1a6a6eda - core::panicking::panic::h1b11496bce256cfc
  │                         at libcore/panicking.rs:52
  │    9:     0x7f0c17aca06c - script::dom::document::Document::maybe_queue_document_completion::h3b1e5b4a090bbabd
  │   10:     0x7f0c17c7f6af - _ZN6script13script_thread12ScriptThread11handle_msgs17h0817e3528e00b258E.llvm.12639552465219462255
  │   11:     0x7f0c17c7abf1 - script::script_thread::ScriptThread::start::h6032fc7704676772
  │   12:     0x7f0c17dee3fe - std::sys_common::backtrace::__rust_begin_short_backtrace::hbc6222a5821321ed
  │   13:     0x7f0c183ac157 - _ZN3std9panicking3try7do_call17h6d9238133ac45be6E.llvm.16677778205194856310
  │   14:     0x7f0c1a698059 - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:102
  │   15:     0x7f0c18136700 - <F as alloc::boxed::FnBox<A>>::call_box::h1ca32cd513fdbbba
  │   16:     0x7f0c1a682cca - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$RP$$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h9d1cd498499ed135
  │                         at liballoc/boxed.rs:682
  │                          - std::sys_common::thread::start_thread::ha4bfe77455e6e434
  │                         at libstd/sys_common/thread.rs:24
  │   17:     0x7f0c1a66f895 - std::sys::unix::thread::Thread::new::thread_start::hebe133da3ea5230a
  │                         at libstd/sys/unix/thread.rs:90
  │   18:     0x7f0c14ea4183 - start_thread
  │   19:     0x7f0c1376b03c - clone
  │   20:                0x0 - <unknown>
  │ [2018-12-22T18:24:02Z ERROR servo] called `Option::unwrap()` on a `None` value
  └ Pipeline failed in hard-fail mode.  Crashing!
```

gterzian committed:
set document ready state to complete in step 4 of parser abort

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/14658.

gterzian commented:
Ok, I've done the following:

1. Turned the panic into a warning(which is consistent to what was already there, by the way, and the previous panic message was incorrect): https://github.com/servo/servo/pull/22121/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fR1062
2. Only checking this "delaying load events mode" if there is actually a browsing context, avoiding the panic: https://github.com/servo/servo/pull/22121/files#diff-30a18e04d7e0b66aafdf192e416cad44R1914
3. Added a new commit for something I came across while debugging the latest, seems like a straightforward change based on the spec.

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit fbfaa16fae777773602d4bce7cd210bc5f561ccb with merge 9357720025e57c8f558e5704d446f82d438be543...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/798), [arm32](http://build.servo.org/builders/arm32/builds/13665), [arm64](http://build.servo.org/builders/arm64/builds/13694), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11252), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11308), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6262), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6238), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10496), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10543), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6301), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6222), [magicleap](http://build.servo.org/builders/magicleap/builds/411), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/F9vNh_12Q9e2_hCsIK5nKQ)
State: approved= try=True

gterzian commented:
Ok, this is what I have done:

1. Turned the panic into a warning(which is consistent to what was already there): https://github.com/servo/servo/pull/22121/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fR1062
2. Only checking this "delaying load events mode" if there is actually a browsing context, avoiding the panic: https://github.com/servo/servo/pull/22121/files#diff-30a18e04d7e0b66aafdf192e416cad44R1914
3. Added a new commit for something I came across while debugging the latest, seems like a straightforward change based on the spec. See step4 of https://html.spec.whatwg.org/multipage/parsing.html#abort-a-parser

gterzian commented:
Added this check for consistency(even though it hadn't crashed at that point yet for lack of a browsing context, like it did in `maybe_queue_document_load`) https://github.com/servo/servo/pull/22121/commits/eb82e781a36c055dab4c1cc3772f7a555abd31a6#diff-1879ac6bd5d567e2aa43529d33474677R2090

paulrouget commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit cca354b has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget cca354b217f4f8c7d92229ec88adde83b8300daa -->

bors-servo commented:
:hourglass: Testing commit cca354b217f4f8c7d92229ec88adde83b8300daa with merge 66223ee1054a36a093cf5ded065f6fc6d3fecd8b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11335)

gterzian commented:
@bors-servo retry

https://github.com/servo/servo/issues/22559 (this one is related, yet really intermittent, cannot reproduce)
https://github.com/servo/servo/issues/22560

bors-servo commented:
:zap: Previous build results for [android-mac](http://build.servo.org/builders/android-mac/builds/820), [arm32](http://build.servo.org/builders/arm32/builds/13687), [arm64](http://build.servo.org/builders/arm64/builds/13716), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11279), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6260), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10517), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10565), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6323), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6243), [magicleap](http://build.servo.org/builders/magicleap/builds/433), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/IneQCQAoSFeFNjL_1ZxZAQ) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11335), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6289)...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/820), [arm32](http://build.servo.org/builders/arm32/builds/13687), [arm64](http://build.servo.org/builders/arm64/builds/13716), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11279), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11338), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6290), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6260), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10517), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10565), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6323), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6243), [magicleap](http://build.servo.org/builders/magicleap/builds/433), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/IneQCQAoSFeFNjL_1ZxZAQ)
Approved by: paulrouget
Pushing 66223ee1054a36a093cf5ded065f6fc6d3fecd8b to master...
